File: src/providers/supabase/rules/fail-fast-env-validation.ts
/**
 * supabase-fail-fast-env-validation (reliability)
 *
 * `createClient(process.env.X, process.env.Y)` with a missing env var throws
 * the SDK's own error ("supabaseUrl is required." / "supabaseKey is
 * required.") — which names the SDK parameter, not YOUR env var. An explicit
 * presence check produces an actionable message naming the exact variable to
 * set (especially confusing otherwise with Next.js NEXT_PUBLIC_* build-time
 * inlining, where the fix is a rebuild, not a restart).
 *
 * Tracks (in source order, since guards precede the call they protect):
 *   - which local names the client factories were imported as
 *     (`createClient` from @supabase/supabase-js, `createBrowserClient` /
 *     `createServerClient` from @supabase/ssr)
 *   - which local variables were assigned directly from `process.env.X`
 *   - which variables/env-vars an `if (!x || ...) throw/return` guard covers
 * then, at the factory call, flags any argument that resolves to a
 * `process.env` value with no matching guard.
 */

function unwrapNonNull(node: any): any {
  return node?.type === 'TSNonNullExpression' ? node.expression : node;
}

function processEnvMemberName(node: any): string | undefined {
  const n = unwrapNonNull(node);
  if (n?.type !== 'MemberExpression' || n.computed) return undefined;
  const obj = n.object;
  if (obj?.type !== 'MemberExpression' || obj.computed) return undefined;
  if (obj.object?.type !== 'Identifier' || obj.object.name !== 'process') return undefined;
  if (obj.property?.type !== 'Identifier' || obj.property.name !== 'env') return undefined;
  if (n.property?.type !== 'Identifier') return undefined;
  return n.property.name;
}

function hasThrowOrReturn(node: any): boolean {
  if (!node) return false;
  if (node.type === 'ThrowStatement' || node.type === 'ReturnStatement') return true;
  if (node.type === 'BlockStatement') {
    return (node.body ?? []).some((s: any) => s.type === 'ThrowStatement' || s.type === 'ReturnStatement');
  }
  return false;
}

const rule = {
  meta: {
    type: 'suggestion',
    docs: {
      description: 'createClient must fail fast when required env vars are missing',
      category: 'reliability',
      rationale:
        'createClient throws immediately when an argument is missing, but with the SDK\'s message ("supabaseKey is required.") — it names the SDK parameter, not your env var. An explicit presence check produces an error naming the exact variable to set, which matters most with Next.js NEXT_PUBLIC_* inlining where the fix is a rebuild with the var present, not a server restart.',
      docsUrl: 'https://supabase.com/docs/reference/javascript/initializing',
      recommended: true,
    },
    messages: {
      missingEnvValidation:
        'createClient is called with {{vars}} with no presence check beforehand. Throw if it/they are unset before calling createClient.',
    },
    schema: [],
  },
  create(context: any) {
    const FACTORY_IMPORTS: Record<string, string[]> = {
      '@supabase/supabase-js': ['createClient'],
      '@supabase/ssr': ['createBrowserClient', 'createServerClient'],
    };
    const factoryLocalNames = new Set<string>();
    const envVarOfVariable = new Map<string, string>();
    const validatedVarNames = new Set<string>();
    const validatedEnvNames = new Set<string>();

    function addTarget(node: any) {
      const n = unwrapNonNull(node);
      if (n?.type === 'Identifier') {
        validatedVarNames.add(n.name);
        return;
      }
      const envName = processEnvMemberName(n);
      if (envName) validatedEnvNames.add(envName);
    }

    function collectGuardTargets(node: any) {
      if (!node) return;
      if (node.type === 'LogicalExpression' && node.operator === '||') {
        collectGuardTargets(node.left);
        collectGuardTargets(node.right);
        return;
      }
      if (node.type === 'UnaryExpression' && node.operator === '!') {
        addTarget(node.argument);
        return;
      }
      if (node.type === 'BinaryExpression' && (node.operator === '==' || node.operator === '===')) {
        const sides = [node.left, node.right];
        const isNullish = (n: any) =>
          (n?.type === 'Literal' && n.value === null) || (n?.type === 'Identifier' && n.name === 'undefined');
        const target = sides.find((s: any) => !isNullish(s));
        const nullSide = sides.find(isNullish);
        if (target && nullSide) addTarget(target);
      }
    }

    return {
      ImportDeclaration(node: any) {
        const factories = FACTORY_IMPORTS[node.source?.value];
        if (!factories) return;
        for (const s of node.specifiers ?? []) {
          if (
            s?.type === 'ImportSpecifier' &&
            s.imported?.type === 'Identifier' &&
            factories.includes(s.imported.name) &&
            s.local?.type === 'Identifier'
          ) {
            factoryLocalNames.add(s.local.name);
          }
        }
      },

      VariableDeclarator(node: any) {
        if (node.id?.type !== 'Identifier') return;
        const envName = processEnvMemberName(node.init);
        if (envName) envVarOfVariable.set(node.id.name, envName);
      },

      IfStatement(node: any) {
        if (!hasThrowOrReturn(node.consequent)) return;
        collectGuardTargets(node.test);
      },

      CallExpression(node: any) {
        if (factoryLocalNames.size === 0) return;
        if (node.callee?.type !== 'Identifier' || !factoryLocalNames.has(node.callee.name)) return;

        const missing: string[] = [];
        for (const rawArg of node.arguments ?? []) {
          const arg = unwrapNonNull(rawArg);
          let envName: string | undefined;
          let isValidated: boolean;

          if (arg?.type === 'Identifier') {
            envName = envVarOfVariable.get(arg.name);
            if (!envName) continue;
            isValidated = validatedVarNames.has(arg.name);
          } else {
            envName = processEnvMemberName(arg);
            if (!envName) continue;
            isValidated = validatedEnvNames.has(envName);
          }

          if (!isValidated) missing.push(envName);
        }

        if (missing.length > 0) {
          context.report({ node, messageId: 'missingEnvValidation', data: { vars: missing.join(', ') } });
        }
      },
    };
  },
};

export const supabaseFailFastEnvValidationRule = rule;
export default rule;

File: src/providers/supabase/rules/python/fail_fast_env_validation.py
"""supabase-fail-fast-env-validation

Parity with the JS rule: `create_client(url, key)` called with values that
trace back to `os.environ` with no presence check first raises the SDK's
own ("supabaseUrl is required." / "supabaseKey is required.") message
instead of one naming the actual env var to set.

Tracks (in source order, since guards precede the call they protect):
  - the local name `create_client` was imported as (`from supabase import
    create_client`)
  - which local variables were assigned directly from an env lookup
    (`os.environ.get("X")` / `os.getenv("X")` / `os.environ["X"]`)
  - which variables/env-names an `if not x: raise/return` (or `is None`)
    guard covers
then, at the factory call, flags any argument that resolves to an
env-sourced value with no matching guard.
"""
from __future__ import annotations

import ast
import sys
from pathlib import Path

_PROVIDER = Path(__file__).resolve().parents[2]
if str(_PROVIDER) not in sys.path:
    sys.path.insert(0, str(_PROVIDER))

RULE_KEY = "supabase-fail-fast-env-validation"


def _env_name(node: ast.AST | None) -> str | None:
    """Returns the env var name for `os.environ.get("X")`, `os.getenv("X")`, or `os.environ["X"]`."""
    if isinstance(node, ast.Call):
        func = node.func
        if isinstance(func, ast.Attribute) and func.attr == "get":
            base = func.value
            if isinstance(base, ast.Attribute) and base.attr == "environ" and _is_os(base.value):
                return _first_str_arg(node)
        if isinstance(func, ast.Attribute) and func.attr == "getenv" and _is_os(func.value):
            return _first_str_arg(node)
        if isinstance(func, ast.Name) and func.id == "getenv":
            return _first_str_arg(node)
        return None
    if isinstance(node, ast.Subscript):
        base = node.value
        if isinstance(base, ast.Attribute) and base.attr == "environ" and _is_os(base.value):
            key = node.slice
            if isinstance(key, ast.Constant) and isinstance(key.value, str):
                return key.value
        return None
    return None


def _is_os(node: ast.AST) -> bool:
    return isinstance(node, ast.Name) and node.id == "os"


def _first_str_arg(call: ast.Call) -> str | None:
    if call.args and isinstance(call.args[0], ast.Constant) and isinstance(call.args[0].value, str):
        return call.args[0].value
    return None


def _has_throw_or_return(stmts: list[ast.stmt]) -> bool:
    return any(isinstance(s, (ast.Raise, ast.Return)) for s in stmts)


def check(tree: ast.AST, path: str, source: str) -> list[dict]:
    factory_local_names: set[str] = set()
    env_var_of_variable: dict[str, str] = {}
    validated_var_names: set[str] = set()
    validated_env_names: set[str] = set()
    out: list[dict] = []

    def add_target(node: ast.AST | None) -> None:
        if isinstance(node, ast.Name):
            validated_var_names.add(node.id)
            return
        env_name = _env_name(node)
        if env_name:
            validated_env_names.add(env_name)

    def collect_guard_targets(node: ast.AST | None) -> None:
        if node is None:
            return
        if isinstance(node, ast.BoolOp) and isinstance(node.op, ast.Or):
            for v in node.values:
                collect_guard_targets(v)
            return
        if isinstance(node, ast.UnaryOp) and isinstance(node.op, ast.Not):
            add_target(node.operand)
            return
        if isinstance(node, ast.Compare) and len(node.ops) == 1:
            op = node.ops[0]
            if isinstance(op, (ast.Is, ast.Eq)):
                sides = [node.left, node.comparators[0]]

                def is_nullish(n: ast.AST) -> bool:
                    return isinstance(n, ast.Constant) and n.value is None

                target = next((s for s in sides if not is_nullish(s)), None)
                null_side = next((s for s in sides if is_nullish(s)), None)
                if target is not None and null_side is not None:
                    add_target(target)

    for node in ast.walk(tree):
        if isinstance(node, ast.ImportFrom) and node.module == "supabase":
            for alias in node.names:
                if alias.name == "create_client":
                    factory_local_names.add(alias.asname or alias.name)

        elif isinstance(node, ast.Assign) and len(node.targets) == 1 and isinstance(node.targets[0], ast.Name):
            env_name = _env_name(node.value)
            if env_name:
                env_var_of_variable[node.targets[0].id] = env_name

        elif isinstance(node, ast.AnnAssign) and isinstance(node.target, ast.Name):
            env_name = _env_name(node.value)
            if env_name:
                env_var_of_variable[node.target.id] = env_name

        elif isinstance(node, ast.If):
            if _has_throw_or_return(node.body):
                collect_guard_targets(node.test)

    if not factory_local_names:
        return out

    for node in ast.walk(tree):
        if not isinstance(node, ast.Call):
            continue
        func = node.func
        if not (isinstance(func, ast.Name) and func.id in factory_local_names):
            continue

        missing: list[str] = []
        for arg in [*node.args, *[kw.value for kw in node.keywords if kw.arg in ("supabase_url", "supabase_key")]]:
            env_name: str | None = None
            is_validated: bool

            if isinstance(arg, ast.Name):
                env_name = env_var_of_variable.get(arg.id)
                if not env_name:
                    continue
                is_validated = arg.id in validated_var_names
            else:
                env_name = _env_name(arg)
                if not env_name:
                    continue
                is_validated = env_name in validated_env_names

            if not is_validated:
                missing.append(env_name)

        if missing:
            line = getattr(node, "lineno", 1) or 1
            col = (getattr(node, "col_offset", 0) or 0) + 1
            out.append({"file": path, "line": line, "column": col, "ruleKey": RULE_KEY})

    return out
